### ISSUE TITLE
undefined reference to `mosquitto_client_id'

### ISSUE THREAD

**Issue by wasawah** [iegomez/mosquitto-go-auth#91]
I'm trying to install it on raspberry pi, but i get this:

```
pi@raspberrypi:~/go/src/mosquitto-go-auth $ make
go build -buildmode=c-archive go-auth.go
go build -buildmode=c-shared -o go-auth.so
# github.com/iegomez/mosquitto-go-auth
/usr/bin/ld: $WORK/b001/_x003.o: in function `mosquitto_auth_acl_check':
auth-plugin.c:(.text+0x3ec): undefined reference to `mosquitto_client_id'
/usr/bin/ld: auth-plugin.c:(.text+0x3f8): undefined reference to `mosquitto_client_username'
collect2: error: ld returned 1 exit status
make: *** [Makefile:3: all] Error 2

```
can you help me out of this, thank you

**Comment by iegomez** [comment]
Hi! As the readme states, when building for Mosquitto 1.5 or later, you need to export some flags before:
```
export CGO_CFLAGS="-I/usr/local/include -fPIC"
export CGO_LDFLAGS="-shared"
make
```
Let me know if that works.

**Comment by wasawah** [comment]
Tried this before, but it doesnt change anything.  Any other advise?

**Comment by iegomez** [comment]
Could you provide more details about your Mosquitto version, how you installed it, etc.?

**Comment by wasawah** [comment]
cloned from git and made requirements

pi@raspberrypi:~/go/src/mosquitto-go-auth $ go version
go version go1.14.6 linux/arm

pi@raspberrypi:~/go/src/mosquitto-go-auth $ go env
GO111MODULE=""
GOARCH="arm"
GOBIN=""
GOCACHE="/home/pi/.cache/go-build"
GOENV="/home/pi/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="arm"
GOHOSTOS="linux"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/pi/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_arm"
GCCGO="gccgo"
GOARM="6"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/pi/go/src/mosquitto-go-auth/go.mod"
CGO_CFLAGS="-I/usr/local/include -fPIC"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-shared"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -marm -pthread -fmessage-length=0
-fdebug-prefix-map=/tmp/go-build262527290=/tmp/go-build
-gno-record-gcc-switches"

**Comment by iegomez** [comment]
I meant Mosquitto: the undefined references come from missing mosquitto headers.

**Comment by wasawah** [comment]
Mosquitto:

wget http://mosquitto.org/files/source/mosquitto-1.6.10.tar.gz
tar zxvf mosquitto-1.6.10.tar.gz
make
sudo make install



V V čet., 6. avg. 2020 ob 20:02 je oseba Ignacio Gómez <
notifications@github.com> napisala:

**Comment by iegomez** [comment]
Please check that headers are at `/usr/local/include`. You're looking for `mosquitto_broker.h`, but `mosquitto.h` and `mosquitto_plugin.h` should also be there.

**Comment by wasawah** [comment]
pi@raspberrypi:/usr/local/include $ ls
libwebsockets    mosquitto_broker.h  mosquittopp.h
libwebsockets.h  mosquitto.h         mqtt_protocol.h
lws_config.h     mosquitto_plugin.h  openssl

they are all here :D

V V čet., 6. avg. 2020 ob 20:32 je oseba Ignacio Gómez <
notifications@github.com> napisala:

**Comment by iegomez** [comment]
That's odd 😕. Could you look into `mosquitto_broker.h` to confirm the conflicting functions are there?

**Comment by wasawah** [comment]
This?


/*
 * Function: mosquitto_client_username
 *
 * Retrieve the username associated with a client.
 */
const char *mosquitto_client_username(const struct mosquitto *client);

file is identical to:
https://github.com/eclipse/mosquitto/blob/master/src/mosquitto_broker.h

V V čet., 6. avg. 2020 ob 22:43 je oseba Ignacio Gómez <
notifications@github.com> napisala:

**Comment by iegomez** [comment]
I don't know what's going on then. Since you're trying in it in a RPi, it may be specific to Raspbian (I'm assuming that's the OS you're running). I haven't tried installing on one since Mosquitto 1.4.x, when those functions didn't exist. So if it's not a Raspbian problem, then I have no other advise besides deleting all and trying to install fresh to see if it was a random hiccup.

**Comment by iegomez** [comment]
Hey, @wasawah, did you have any luck after reporting the issue? Is there something I can do to help you?
If you did get it to work, could you share any results and then close the issue?

Cheers!

**Comment by arctic-alpaca** [comment]
I see the same issue trying to build a docker container on Rasperry Pi using the supplied Dockerfile after replacing `.linux-amd64.` with `.linux-armv6l`.

**Comment by iegomez** [comment]
> Using Mosquitto version 1.4.15, I was able to create a container. When trying to load it, mosquitto quits with `Error: Incorrect auth plugin version (got 4, expected 2).` This seems to be related to libwebsockets as seen [here](https://github.com/jpmens/mosquitto-auth-plug/issues/353).

As per this recommendation from Roger in https://github.com/iegomez/mosquitto-go-auth/issues/96, the latest version (1.1.0) of the plugin returns 4 to support both current and upcoming version of Mosquitto. I don't know the details on why the older function won't work anymore, so I can only recommend trying with Mosquitto's latest version or using an older release of the plugin (1.0.0 should work, unless you need MySQL root cert which was also fixed in 1.1.0).

**Comment by arctic-alpaca** [comment]
Creating a container with plugin v.1.0.0 and Mosquitto 1.4.15 seems to work, thanks! 

I can't get the plugin compiled with a >=1.5 version of Mosquitto in any way. Every attempt results in the previously stated errors. 
Do you have any thoughts about where the problem could be? I'm not sure whether this is an arm specific issue. This is pure speculation, maybe the cgo flags aren't processed correctly on arm since adding those flags allowed proper compilation on other architectures.

**Comment by iegomez** [comment]
@arctic-alpaca I'm glad it worked! I'm not sure what might be going on with the RPi since I haven tried in one since 1.4, maybe may be related to cross compilation? BTW, have you ever tried building mosquitto >= 1.5 + plugin directly in the RPi?

Anyway, I found these resources in cse they may be of any help:
- https://medium.com/@chrischdi/cross-compiling-go-for-raspberry-pi-dc09892dc745
- https://gist.github.com/deepakkamesh/0f4667c2ea8bc02309a6ae75b2181ac0

**Comment by arctic-alpaca** [comment]
I have tried cross compilation with docker buildx and creating the docker container directly on the RPI, both approaches lead to the described error with version 1.6.12 of Mosquitto and version 1.1.0 of this plugin.

The cross compilation done by docker buildx should be relatively similar to native compilation on arm.

**Comment by iegomez** [comment]
@arctic-alpaca Sorry I can't help much there, I don't have means to try this. On another topic, regarding you having to go back to release 1.0.0 to compile against Mosquitto 1.4.x, I'll try to add an optional build so you may use latest version. Will keep you posted.

**Comment by arctic-alpaca** [comment]
> On another topic, regarding you having to go back to release 1.0.0 to compile against Mosquitto 1.4.x, I'll try to add an optional build so you may use latest version. Will keep you posted.

That would be great, thank you!

> Sorry I can't help much there, I don't have means to try this.

I'm out of ideas what to try here. Since using Mosquitto 1.4.x currently works for me, I'll leave it at that. If I find any new information, I'll add it here.

**Comment by arctic-alpaca** [comment]
I thought I'd add a description on how I created the working docker container and how the error can be reproduced:
I'm using Windows 10 as host and Debian buster inside virtual box, but this should work on any debian amd64 system.

1. install docker https://docs.docker.com/engine/install/debian/#install-using-the-repository
2. make buildx availabe `export DOCKER_CLI_EXPERIMENTAL=enabled`
3. add user to docker group `sudo usermod -aG docker $(whoami)`
4. relog into user, `su` then `su user` for example
5. install qemu `sudo apt install -y qemu qemu-user-static qemu-user binfmt-support`
6. configure docker 
    ```
   docker buildx create --name=multi
   docker buildx inspect --bootstrap multi
   docker buildx use multi
    ```
7. Build and push to registry `docker buildx build --output type=image,name="repository/applicationname",push=true --platform linux/arm/v7 --tag "repository/applicationname:version" --file Dockerfile .`

This is the dockerfile I used. It could probably be optimized quite a bit, but this is what I got working with Mosquitto 1.4.x and Mosquitto-Go-Auth 1.0.0. The error produced by Mosquitto >=1.5.x can also be reproduced like this.
```dockerfile
FROM debian:10.6 AS builder1
# Set Mosquitto version
#ENV VERSION=1.4.15 
ENV VERSION=1.6.12 



RUN apt-get update && apt-get install -y  libc-ares2 libc-ares-dev openssl uuid uuid-dev wget build-essential git cmake make libwebsockets8 libwebsockets-dev

RUN wget https://mosquitto.org/files/source/mosquitto-${VERSION}.tar.gz -O /tmp/mosq.tar.gz
RUN mkdir -p /build/mosq
RUN tar --strip=1 -xf /tmp/mosq.tar.gz -C /build/mosq
RUN rm /tmp/mosq.tar.gz
RUN make -C /build/mosq -j "$(nproc)" \
        CFLAGS="-Wall -O2 -I/build/lws/include" \
        LDFLAGS="-L/build/lws/lib" \
        WITH_ADNS=no \
        WITH_DOCS=no \
        #WITH_SHARED_LIBRARIES=yes \
        WITH_SRV=no \
        #WITH_STRIP=no \
        WITH_WEBSOCKETS=yes \
        prefix=/usr \
        binary
RUN mkdir -p /mosquitto/config /mosquitto/data /mosquitto/log
RUN install -d /usr/sbin/
RUN install -s -m755 /build/mosq/client/mosquitto_pub /usr/bin/mosquitto_pub 
RUN install -s -m755 /build/mosq/client/mosquitto_sub /usr/bin/mosquitto_sub 
RUN install -s -m644 /build/mosq/lib/libmosquitto.so.1 /usr/lib/libmosquitto.so.1 
RUN install -s -m755 /build/mosq/src/mosquitto /usr/sbin/mosquitto 
RUN install -s -m755 /build/mosq/src/mosquitto_passwd /usr/bin/mosquitto_passwd 
RUN install -m644 /build/mosq/mosquitto.conf /mosquitto/config/mosquitto.conf



#Use debian:stable-slim as a builder and then copy everything.
#FROM debian:stable-slim as builder

FROM --platform=$BUILDPLATFORM golang:1.14.10-buster AS builder2

#Set plugin version
ENV PLUGIN_VERSION=1.0.0

ENV CGO_ENABLED=1
ENV GOOS=linux
ENV GOARCH=arm
ENV GOARM=7
ENV PKG_CONFIG_PATH=/usr/lib/arm-linux-gnueabihf/pkgconfig
ENV PKG_CONFIG_LIBDIR="${PKG_CONFIG_LIBDIR}:/usr/local/lib:/usr/local/include"\
ENV CC=arm-linux-gnueabihf-gcc
ENV CGO_CFLAGS="-I/usr/local/include -fPIC"
ENV CGO_LDFLAGS="-shared"

# Install TARGETPLATFORM parser to translate its value to GOOS, GOARCH, and GOARM
COPY --from=tonistiigi/xx:golang / /
# Bring TARGETPLATFORM to the build scope
ARG TARGETPLATFORM


WORKDIR /app

RUN dpkg --add-architecture armv7

RUN apt-get update && apt-get install -y libwebsockets8 libwebsockets-dev libc-ares2 libc-ares-dev openssl uuid uuid-dev wget build-essential git gcc-arm-linux-gnueabihf libc6-dev-armhf-cross

COPY --from=builder1 /build/mosq/src/mosquitto_broker.h /usr/local/include/mosquitto_broker.h
COPY --from=builder1 /build/mosq/src/ /usr/local/include/
COPY --from=builder1 /build/mosq/src/mosquitto_plugin.h /usr/local/include/mosquitto_plugin.h
COPY --from=builder1 /build/mosq/lib/ /usr/local/include/
COPY --from=builder1 /build/mosq/lib/ /usr/local/lib/
COPY --from=builder1 /mosquitto/ /app/mosquitto/
COPY --from=builder1 /build/mosq/src/mosquitto /usr/sbin/mosquitto


WORKDIR /app
#Get / build the plugin.
RUN mkdir mosquitto-go-auth
RUN cd mosquitto-go-auth && \
   git clone --branch $PLUGIN_VERSION https://github.com/iegomez/mosquitto-go-auth.git . && \
   export PATH=$PATH:/usr/local/go/bin && \
   export GOPATH=$GOPATH:/app/mosquitto-go-auth && \
   export CGO_CFLAGS="-I/usr/local/include -fPIC"  && \
   export CGO_LDFLAGS="-shared -L/usr/local/include -fPIC" && \
   #make
   go build -buildmode=c-archive /app/mosquitto-go-auth/go-auth.go   && \
   go build -buildmode=c-shared -o /app/mosquitto-go-auth/go-auth.so  && \
   go build pw-gen/pw.go 





#Start from a new image.
FROM debian:10.6-slim

#Get mosquitto dependencies.
RUN apt update && apt install -y libwebsockets8 libc-ares2 openssl uuid
#RUN apk --no-cache add ca-certificates libressl

#Setup mosquitto env.
RUN mkdir -p /var/lib/mosquitto /var/log/mosquitto
RUN groupadd mosquitto \
    && useradd -s /sbin/nologin mosquitto -g mosquitto -d /var/lib/mosquitto \
    && chown -R mosquitto:mosquitto /var/log/mosquitto/ \
    && chown -R mosquitto:mosquitto /var/lib/mosquitto/

#Copy confs, plugin so and mosquitto binary.
COPY --from=builder2 /app/mosquitto/ /mosquitto/
COPY --from=builder2 /app/mosquitto-go-auth/pw /mosquitto/pw
COPY --from=builder2 /app/mosquitto-go-auth/go-auth.so /mosquitto/go-auth.so
COPY --from=builder2 /usr/sbin/mosquitto /usr/sbin/mosquitto


#Expose tcp and websocket ports as defined at mosquitto.conf (change accordingly).
EXPOSE 1883 1883

ENTRYPOINT ["sh", "-c", "/usr/sbin/mosquitto -c /etc/mosquitto/mosquitto.conf" ]
#CMD echo "Hello world" 
```

**Comment by iegomez** [comment]
@arctic-alpaca Thanks for sharing that description, it'll surely be useful for anyone following your approach!

On the plugin version issue, I just created this PR which I believe fixes it: https://github.com/iegomez/mosquitto-go-auth/pull/103
Nevertheless, I'm waiting on comments in https://github.com/iegomez/mosquitto-go-auth/issues/96 to see if thereś anything else that needs consideration.

Cheers!

**Comment by iegomez** [comment]
@arctic-alpaca I already merged cache and mosquitto version changes and released them as 1.2.0, so let me know if it works after giving it a try,

**Comment by arctic-alpaca** [comment]
Thank you very much! Compiling and using v1.2.0 with Mosquitto 1.4.15 now works.

**Comment by tercelcn** [comment]
I use ubuntu 18.04, go 1.15.3 , mosquitto 1.6.12, linux-amd64 is ok.
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/john/.cache/go-build"
GOENV="/home/john/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/john/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/john/go"
GOPRIVATE=""
GOPROXY="https://goproxy.io,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/john/Downloads/mosquitto-go-auth-1.2.0/go.mod"
CGO_CFLAGS="-fPIC -I/home/john/Downloads/linux_amd64/usr/local/include"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-shared -L/home/john/Downloads/linux_amd64/usr/local/lib"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build459617400=/tmp/go-build -gno-record-gcc-switches"

**Comment by iegomez** [comment]
Sorry, @tercelcn, it's a bit hard to parse that. How did you try to build the plugin and what version of it, or was it the main branch?

**Comment by mkohns** [comment]
Hi, I stumbled across this issue and debugged a bit. I still do not know the root cause for mosquitto-go-auth no building on raspberry pi - but i found a dirty solution for building this setup for rpi4

ENV MOSQUITTO_VERSION=1.6.10
ENV GO_VERSION=1.13.8

The reported linker error from @arctic-alpaca can actually be ignored. The reference is present - no idea why cgo has trouble finding it. 
With this patch file saved to patch.go-auth.go.rpi4 in docker subdirectoy

```
2a3
> // #cgo linux armhf LDFLAGS: -Wl,-unresolved-symbols=ignore-all
```

and this change in Dockerfile

```
#Copy the patch for rpi
COPY patch.go-auth.go.rpi4 .

#Get / build the plugin.
RUN mkdir mosquitto-go-auth && \
    cd mosquitto-go-auth && \
    git clone https://github.com/iegomez/mosquitto-go-auth.git . && \
    export PATH=$PATH:/usr/local/go/bin && \
    export CGO_CFLAGS="-I/usr/local/include -fPIC" && \
    export CGO_LDFLAGS="-shared" && \
    patch go-auth.go ../patch.go-auth.go.rpi4 && \
    make
```
The docker image can be build and runs smoothly with the newest version of mosquitto and auth plugin.

I suppose the patch file can be merged into master as the cgo directive should only hit armhf. But please double check.

**Comment by tercelcn** [comment]
@mkohns Thanks!!!   
My env is ubuntu 18.04, go 1.15.3 ,mosquitto 1.6.12 and go auth 1.1. I can successfully build linux/arm when I modify 
`export  CGO_LDFLAGS="-shared -Wl,-unresolved-symbols=ignore-all"`  
I modify "go build  -x -buildmode=c-shared -o go-auth.so", I saw that cgo attach "-pie" to link parameters when env is   linux/arm.

**Comment by mrdc** [comment]
@iegomez @tercelcn 
I can confirm that tercelcn’s method works fine: I’ve successfully complied the latest mosquitto-go-auth with go1.15.5 on RPi3b+ running Raspbian Buster (10). Also the latest mosquitto 1.6.12.
Will test later if it works as expected.

**Comment by iegomez** [comment]
Nice! Let's fix it during the week so it' generally available. Thanks @mkohns for figurin it out!

**Comment by wasawah** [comment]
Got new error, after building as @tercelcn suggested

pi@raspberrypi:~/mqtt/mosquitto-1.6.10 $ mosquitto -c mosquitto-auth.conf
1607254607: mosquitto version 1.6.10 starting
1607254607: Config loaded from mosquitto-auth.conf.
1607254607: Loading plugin: /home/pi/mqtt/mosquitto-go-auth/go-auth.so
1607254607: ├── Username/password checking enabled.
1607254607: ├── TLS-PSK checking enabled.
1607254607: └── Extended authentication not enabled.
FATA[2020-12-06T11:36:47Z] backends error

**Comment by iegomez** [comment]
Hello, everyone! Anyone willing to add instructions to the docs to get it working in RPi4? If so, just open a PR and I'll be happy to review it.

Cheers!

**Comment by mrdc** [comment]
@iegomez 

Hi,

It again fails to compile on RPi with `mosquitto_client_id` error :(

`export CGO_LDFLAGS="-shared -Wl,-unresolved-symbols=ignore-all"` doesn’t help.
Probably I also did some extra steps previous time - don’t remember.

```
mosquitto-go-auth 1.5.0
mosquito 2.0.10 (from source)
go 1.6.3
```

**Comment by arctic-alpaca** [comment]
You can try [this Dockerfile](https://github.com/iegomez/mosquitto-go-auth/blob/023be3eaa41f92a25ad2dca202a433ad4a5de2d9/Dockerfile).
Or revert to a mosquitto version before 2, I think there were some extra steps with docker and version 2.x

**Comment by mrdc** [comment]
@arctic-alpaca 

Unfortunately, I'm not using Docker for mosquitto.

**Comment by mrdc** [comment]
Success!

This `Makefile` fixes the issue (mosquitto 2.0.10 and mosquitto-go-auth 1.5.0) . Somehow `export CGO_LDFLAGS="-shared -Wl,-unresolved-symbols=ignore-all"` doesn't work, but changes to `Makefile` work.

```
CFLAGS := -I/usr/local/include -fPIC
LDFLAGS := -shared -Wl,-unresolved-symbols=ignore-all

UNAME_S := $(shell uname -s)

ifeq ($(UNAME_S),Darwin)
        LDFLAGS += -undefined dynamic_lookup
endif

all:
        @echo "Bulding for $(UNAME_S)"
        env CGO_CFLAGS="$(CFLAGS)" go build -buildmode=c-archive go-auth.go
        env CGO_LDFLAGS="$(LDFLAGS)" go build -x -buildmode=c-shared -o go-auth$
        go build pw-gen/pw.go
```
